### ISSUE TITLE
FlutterJNI.loadLibrary called more than once

### ISSUE THREAD

**Issue by penhorwood** [firebase/flutterfire#9964]
## Bug report

**Describe the bug**
The FlutterJNI.loadLibrary is called twice.  

Error message displayed in the Debug Console
**W/FlutterJNI(10613): FlutterJNI.loadLibrary called more than once
W/FlutterJNI(10613): FlutterJNI.prefetchDefaultFontManager called more than once**

### Steps to reproduce

1. Start a simple Flutter App with FirebaseMessaging
2. Use the FirebaseMessaging.onBackgroundMessage( backgroundRouter );
3. Run the App in Debug or Release mode

### Expected behavior

The FlutterJNI.loadLibrary will only be called once.

### Sample project
https://github.com/penhorwood/firetwice
---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[√] Flutter (Channel stable, 3.3.8, on Microsoft Windows [Version 10.0.19044.2251], locale en-US)
[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[√] Android Studio (version 2021.2)
[√] Connected device (1 available)
[√] HTTP Host Availability

```

</details>

---

---

**Comment by darshankawar** [comment]
Closing as duplicate of https://github.com/firebase/flutterfire/issues/9912. Please upgrade plugin version to latest to see the fix.

**Comment by penhorwood** [comment]
This is not a duplicate as this is still happening

I have all of the latest fixes but the duplicate Flutter.JNI.loadLibrary is still loading twice

**Comment by penhorwood** [comment]
Here are two screenshots of my app loading today - 11/24/2022
One is with the **onBackgroundMessage** used the other with that line commented out.  You can see that make the difference in whether **Flutter.JNI.loadLibrary** loads twice.

![image](https://user-images.githubusercontent.com/1919238/203785822-5cc5d9d3-26e5-471f-a4c6-d41bc4ea7794.png)

2nd without **onBackgroundMessage**

![image](https://user-images.githubusercontent.com/1919238/203785890-b5a8be68-a704-4441-aed0-f07a0112db5e.png)

**Comment by darshankawar** [comment]
Thanks for the feedback @penhorwood 
Can you please provide a complete minimal reproducible code sample in text format that shows the behavior ?

**Comment by penhorwood** [comment]
I have a complete example here https://github.com/penhorwood/firetwice.  That is my minimal reproducible code sample.  If you want me to post the files here I can.

**Comment by darshankawar** [comment]
Thanks for the update. I looked at the repo you provided above and observed that the background handler is not the topmost method, unlike in the plugin example.

https://github.com/firebase/flutterfire/blob/26087bb2d1747eb5db82c98424b6d63071137529/packages/firebase_messaging/firebase_messaging/example/lib/main.dart#L47

Also, try to have the `@pragma('vm:entry-point')` annotation as mentioned in the comments above the declaration and see if it helps in your case.

**Comment by penhorwood** [comment]
I followed your instructions.  No change on the results.

W/FlutterJNI( 5559): FlutterJNI.loadLibrary called more than once
W/FlutterJNI( 5559): FlutterJNI.prefetchDefaultFontManager called more than once

You can view my changes on the repo.

**Comment by penhorwood** [comment]
Just an FYI.  I think when it says top most method it is talking about as in not part of a class.  Where it is in main.dart should not matter but that is just my current understanding.  I could be wrong.

**Comment by darshankawar** [comment]
Thanks for the update. Using the sample repo, I do see the same behavior as reported. 
Keeping it open for team's attention and insights.

**Comment by Lyokone** [comment]
Thanks for the report, I'm investigating this

**Comment by mark8044** [comment]
Seeing the same thing here

**Comment by sadaqatdev** [comment]
D/FlutterGeolocator(29054): Creating service.
D/FlutterGeolocator(29054): Binding to location service.
D/FlutterLocationService(29054): Creating service.
D/FlutterLocationService(29054): Binding to location service.
D/FlutterGeolocator(29054): Flutter engine connected. Connected engine count 1
W/FlutterJNI(29054): FlutterJNI.loadLibrary called more than once
W/FlutterJNI(29054): FlutterJNI.prefetchDefaultFontManager called more than once
W/FlutterJNI(29054): FlutterJNI.init called more than once
D/FlutterGeolocator(29054): Flutter engine connected. Connected engine count 2

**Comment by pushpendraKh** [comment]
I am also facing the same issue. @darshankawar Any progress on this?

**Comment by eudorolshop** [comment]
this issue happened to me.
occurs when the application is first run only and in the app foreground.

after receiving/displaying notification in foreground then after a few seconds the app got crashed (force close/Wait & Ok Prompt).
if i choose Wait, the app and notification (fcm) running smoothly.

I've tried various ways to overcome this issue, but no significant result.

please, any workaround for this issue?


library use :

firebase_messaging: 14.2.2
firebase_core: 2.5.0

flutter doctor :

```
[√] Flutter (Channel stable, 3.7.3, on Microsoft Windows [Version 10.0.18363.592], locale en-US)
[√] Windows Version (Installed version of Windows is version 10 or higher)
[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[√] Chrome - develop for the web
```

Error Log : `Attempted to start a duplicate background isolate. Returning`

**Comment by danielgomezrico** [comment]
@eudorolshop that seems like a different issue, WDYT?

**Comment by eudorolshop** [comment]
> @eudorolshop that seems like a different issue, WDYT?
> 
> @LyokoneIt is still happening to me on:
> 
> ```yaml
>   firebase_core: 2.4.1
>   firebase_crashlytics: 3.0.11
>   firebase_analytics: 10.1.0
>   firebase_messaging: 14.2.1
> ```
> 
> And on:
> 
> ```yaml
>   firebase_core: 2.5.0
>   firebase_crashlytics: 3.0.12
>   firebase_analytics: 10.1.1
>   firebase_messaging: 14.2.2
> ```

@danielgomezrico i got this issue too, FlutterJNI.loadlibrary called more than once.

**Comment by sadaqatdev** [comment]
this cause by location plugin

**Comment by cindyha** [comment]
It's still happening to me on 
`firebase_messaging: ^14.2.4
  firebase_core: ^2.6.1`

![firebase_messaging_issue](https://user-images.githubusercontent.com/5300207/220317556-b9a3520e-28e4-41d6-b321-51491c606ae0.png)

**Comment by lybur** [comment]
The same happens to me when the app is in its first launch.
I'm using firebase_messaging: ^14.1.1
![image](https://user-images.githubusercontent.com/110459982/220537247-dd940188-d19b-4af5-97d6-0fc7a1fe413f.png)

**Comment by danielgomezrico** [comment]
Maybe it is related to this issue?
- https://github.com/firebase/flutterfire/issues/9689

**Comment by penhorwood** [comment]
Any movement on this bug?  This is still happening to me with firebase_messaging: 14.5.0

**Comment by juliavi** [comment]
Same here. Any updates on this issue?

**Comment by abdalmonem** [comment]
@darshankawar 

This issue is serous danger!
I investigate it , and i come with:
1 - the main isolate called one time in the background
2 - the main isolate thats triggered from the "FCM" going to load the plugins and act normally as a UI isolate, and keep trying render widgets.
3 - when "main" get trigger by FCM notification the "onBackgroundMessage" going to re register it self with another one, (here we got the error ) , and this is why you got just one.
4 - even if "onBackgroundMessage" handler is not crashed in first time , it will going to call it self in a recursion way 

the normal behave :
The "FCM" trigger the main isolate but should skip at least run "runapp()", this going to solve the issue if we call "onBackgroundMessage" inside runApp and not directly in the main.

Suggested solution:
provide a way to check which "VM entry point" has called the main.
this will provide way to avoid re initiate  "onBackgroundMessage" and give the developers options to deal logically on they way.

**Comment by yeikel16** [comment]
The same happens to me when the app is in its first launch.

-  firebase_core: ^2.15.0
- firebase_messaging: ^14.6.5
- firebase_crashlytics: ^3.3.4

<img width="1082" alt="Captura de pantalla 2023-08-01 a la(s) 3 20 31 p m" src="https://github.com/firebase/flutterfire/assets/26438532/7582605a-b7ba-460a-bc94-4c3190fc9a3f">

**Comment by abdalmonem** [comment]
@yeikel16 i believe we can skip this issue if Dart team add any way to find the name of `entry-point`that ran the main

**Comment by milindgoel15** [comment]
It happens with my app and I am not using firebase messaging. 
The packages I am using are firebase core, performance and analytics. So I don't think it's an issue on messaging. Probably related to core

**Comment by nukeolay** [comment]
Have same issue. Any success with solving it?

**Comment by egedib** [comment]
> It happens with my app and I am not using firebase messaging. The packages I am using are firebase core, performance and analytics. So I don't think it's an issue on messaging. Probably related to core

That's the separate issue - this issue is for the problem related to onBackgroundMessage - check the opening comment. I suggest creating a separate issue, but check StackOverflow first.

> I also faced this issue. When I get **duplicated push notification** in background state then that log is found.

I suspect this is also a different issue - I have this problem with onBackgroundMessage, but the notifications are arriving only once.

**Comment by aly22** [comment]
Was facing the same problem, After writing  `@pragma('vm:entry-point')` before the top level function it fixed it.
Like this:
```
@pragma('vm:entry-point')
Future<void> _firebaseBackgroundMessage(RemoteMessage message) async {
  if (message.notification != null) {
    print('A message received');
  }
}

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  SystemChrome.setEnabledSystemUIMode(SystemUiMode.manual, overlays: [
    SystemUiOverlay.top,
    SystemUiOverlay.bottom,
  ]);
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );
  await PushNotifications.init();

  runApp(const MyApp());
  FirebaseMessaging.onBackgroundMessage(_firebaseBackgroundMessage);
}
```

**Comment by penhorwood** [comment]
I tried the **@pragma('vm:entry-point')** above my methods.  It made no difference.

**Comment by gotiobg** [comment]
This issue was resolved in my case when I removed "just_audio_background" package from pubspec.yaml. If you don't have this package, go through other packages you have and check if any of them invoke a call to main(). This can be done by removing one package at a time and running the app to check if the invocation occurs.

**Comment by xwarzdev** [comment]
Any update? Same problem here 2 main apps running in callstack, tryed the pragma anotation but didnt work.
If I dont call  FirebaseMessaging.onBackgroundMessage( _handler ); only 1 main app running in callstack.

 firebase_messaging: ^14.7.4
 firebase_core: ^2.22.0

**Comment by micant** [comment]
> If someone can follow the steps I describe above and test what I did works, will be apreciated.

I have the same configuration as you. 

If I register `FirebaseMessaging.onBackgroundMessage(_handler)`, I get two calls of main in my call stack and I get some unexpected behaviour, where routing via context works some of the time. If I remove the handler, everything works as expected.

**Comment by gabrielgiovans** [comment]
I am having the same issue when using the FirebaseMessaging.onBackgroundMessage(_handler) function.

FirebaseMessaging Version: 14.7.9

Any updates on this bug?
I see threads with dates of 2 years or more and no solution...

**Comment by Mr-silentcoder** [comment]
if(Firebase.apps.isEmpty){
   await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );
}

**Comment by corepuncher** [comment]
Same issue here. I made sure to use a single instance of FirebaseMessaging, but did not help. Seems like an annoyance more than anything.
